device/alpaka/include/traccc/alpaka/utils/definitions.hpp

```diff
@@ -11,15 +11,42 @@
 #include <alpaka/example/ExampleDefaultAcc.hpp>
 
 #ifdef ALPAKA_ACC_GPU_CUDA_ENABLED
-#include "vecmem/utils/cuda/copy.h"
+#include <vecmem/utils/cuda/copy.hpp>
 #endif
 
-#include "vecmem/utils/copy.hpp"
+#include <vecmem/utils/copy.hpp>
 
 namespace traccc::alpaka {
 
+#ifdef ALPAKA_ACC_GPU_CUDA_ENABLED
+#define WARP_SIZE 32
+#else
+#define WARP_SIZE 4
+#endif
```

Comment:
What is this? :confused: I don't see the `WARP_SIZE` macro used anywhere else in this PR. :confused:

Is `4` something that you would expect that a CPU would have as number of threads? Or where does that value come from?

Even if we decide to make this choice at compile time, which I'm not convinced that we should (it would be better to make this decision at runtime if possible), we should not use a pre-processor macro for this. Rather, I would do something like:

```c++
static constexpr std::size_t WARP_SIZE =
#ifdef ALPAKA_ACC_GPU_CUDA_ENABLED
    32
#else
    4
#endif
    ;
```

Reply:
Work group sizes are hardcoded in a number of places at the moment. That is generally not a terrible idea, since the ideal work group size is a bit more dependent on the algorithm than on the underlying hardware. (Though it's not a black or white distinction.)

But when it comes to figuring out how many threads would be running in parallel in a work group, so the actual "warp size", there we should avoid hardcoding. For consumer GPUs, and all NVIDIA GPUs, this number is `32` of course. But for instance AMD data center GPUs have a "wavefront width" of 64. (Which, if we don't account for it, can mean a 2x performance loss for our code right away.)

Also keep in mind that CPUs (at least through OpenCL) report their capabilities like:

```
  Platform Name                                   Intel(R) OpenCL
Number of devices                                 1
  Device Name                                     Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz
  Device Vendor                                   Intel(R) Corporation
  Device Vendor ID                                0x8086
  Device Version                                  OpenCL 3.0 (Build 0)
  Driver Version                                  2023.15.3.0.20_160000
  Device OpenCL C Version                         OpenCL C 3.0 
  Device Type                                     CPU
  Device Profile                                  FULL_PROFILE
  Device Available                                Yes
  Compiler Available                              Yes
  Linker Available                                Yes
  Max compute units                               16
  Max clock frequency                             3600MHz
  Device Partition                                (core)
    Max number of sub-devices                     16
    Supported partition types                     by counts, equally, by names (Intel)
    Supported affinity domains                    (n/a)
  Max work item dimensions                        3
  Max work item sizes                             8192x8192x8192
  Max work group size                             8192
  Preferred work group size multiple              128
  Max sub-groups per work group                   2048
  Sub-group sizes (Intel)                         4, 8, 16, 32, 64
  Preferred / native vector sizes                 
    char                                                 1 / 32      
    short                                                1 / 16      
    int                                                  1 / 8       
    long                                                 1 / 4       
    half                                                 0 / 0        (n/a)
    float                                                1 / 8       
    double                                               1 / 4        (cl_khr_fp64)
  Half-precision Floating-point support           (n/a)
...
```

I.e. this 8 core / 16 thread CPU will happily tell you to use up to `8192` threads per block if you wish. :stuck_out_tongue: So we'll need to think a bit about how we'd want to run these algorithms on CPUs exactly. (The final answer will probably just be that we wouldn't... :stuck_out_tongue:)